### ISSUE TITLE
Question: Proper auth module usage

### ISSUE THREAD

**Issue by fstr** [terraform-google-modules/terraform-google-kubernetes-engine#849]
I'm currently using the helm provider (v2+) and the auth submodule to get credentials for my private GKE cluster as shown on in the [module readme](https://github.com/terraform-google-modules/terraform-google-kubernetes-engine/tree/master/modules/auth#kuberneteshelm-provider-output). This works perfectly fine for the first `terraform apply` run.

This approach is also describe in the [Google provider docs](https://registry.terraform.io/providers/hashicorp/google/latest/docs/guides/using_gke_with_terraform#using-the-kubernetes-and-helm-providers) and similarly but not identical in the [Kubernetes provider docs](https://registry.terraform.io/providers/hashicorp/kubernetes/latest/docs#statically-defined-credentials).

```
module "gke_auth" {
  source = "terraform-google-modules/kubernetes-engine/google//modules/auth"

  project_id           = var.project_id
  cluster_name         = var.gke_cluster_name
  location             = var.gke_cluster_location
  use_private_endpoint = true
}

provider "helm" {
  kubernetes {
    host                   = module.gke_auth.host
    cluster_ca_certificate = module.gke_auth.cluster_ca_certificate
    token                  = module.gke_auth.token
  }
}

resource "helm_release" { ... }
```

If I return more than one hour later and execute `terraform destroy`, my terraform config does not generate a new token. The previous token is re-used by Terraform, but that token was only valid for 60 minutes.

After running `terraform destroy` more than 60 minutes later:
```
Error: Kubernetes cluster unreachable: invalid configuration: no configuration has been provided, try setting KUBERNETES_MASTER environment variable
```

This is also described in the [Kubernetes provider manual exec-based-credential-plugins](https://registry.terraform.io/providers/hashicorp/kubernetes/latest/docs#exec-based-credential-plugins):
> Some cloud providers have short-lived authentication tokens that can expire relatively quickly ...

I have tried to build the equivalent of the eks exec-based-credential example, but that didn't work for me.

> For further reading, see these examples which demonstrate different approaches to keeping the cluster credentials up to date: AKS, EKS, and GKE.
The [further reading GKE example](https://github.com/hashicorp/terraform-provider-kubernetes/tree/master/_examples/gke#kubeconfig-for-manual-cli-access) mentions the issue of short lived tokens once more, but also claims that a new kubeconfig should be created if another `terraform apply` is executed.

The only thing that works for me, is to manually run (or maybe as a null_resource) `gcloud container clusters get-credentials my-cluster --internal-ip` and then reference the generated kube config in my helm provider block.

```
provider "helm" {
  kubernetes {
    config_path = "/path/to/config"
  }
}
```

I'm not sure if this is a bug, misconfiguration or misunderstanding. How can the auth module be used for multiple `terraform apply` and/or `terraform destroy` runs?

**Comment by morgante** [comment]
It sounds like `data.google_client_config` is somehow picking up a stale access token but I haven't seen this before. What credentials are you using for the provider? (User credentials? Default credentials on a VM? A service account key?)

@rileykarson @bharathkkb Have either of you seen this?

**Comment by fstr** [comment]
I'm using:
```
terraform 0.14.7
hashicorp/google 3.58.0
hashicorp/helm 2.0.2
terraform-google-modules/kubernetes-engine/google/14.0.1
```

I execute Terraform on a local machine using a **service account key**:
```
# Assume /path/to/key.json contains the SA key for tfsa@myproject.iam.gserviceaccount.com
export GOOGLE_APPLICATION_CREDENTIALS="/path/to/key.json"
gcloud auth activate-service-account --key-file="/path/to/key.json"
# returns valid token
gcloud auth describe tfsa@myproject.iam.gserviceaccount.com
# kubectl works too
gcloud container clusters get-credentials mycluster --internal-ip
kubectl get nodes
# time to run terraform
terraform destroy -auto-approve

helm_release.app: Destroying... [id=my-app]
Error: Kubernetes cluster unreachable: invalid configuration: no configuration has been provided, try setting KUBERNETES_MASTER environment variable
```

**Comment by morgante** [comment]
If you run `terraform refresh` does it fix things?

**Comment by bharathkkb** [comment]
When running `destroy` does it show an entry in the logs that the token is being refreshed?
```
....data.google_client_config.default: Refreshing state...
```

**Comment by rileykarson** [comment]
I applied a config w/ the token datasource a few minutes ago, let me see if I get a stale token in an hour. It seems like I'm getting the same token back over and over again from my tokensource, so I think it's cached. I would expect it to use the creds to generate a new token when it expires, though.

**Comment by fstr** [comment]
When running `destroy` there was no sign of refreshing state for the `data.google_client_config.default` data source.

> If you run terraform refresh does it fix things?

This did actually fix things. After running a manual `terraform refresh`, the follow-up `terraform destroy` run passed successfully. 

I find this quite interesting, since `terraform destroy --help` lists `-refresh=true`.
>  -refresh=true          
> Update state prior to checking for differences. This has no effect if a plan file is given to apply.

**Comment by bharathkkb** [comment]
@fstr @rileykarson looks like it was an upstream issue of not doing refresh during destroy and will be fixed in 0.15
https://github.com/hashicorp/terraform/pull/27408

**Comment by fstr** [comment]
Thanks for all your help! I did a terraform `apply` run with terraform `v0.15.0-beta1` which contains the patch from hashicorp/terraform#27408. I waited for more than 60 minutes and was able to successfully run terraform `destroy` without the need of a manual terraform refresh in-between.

Confirmed that it works again in TF 0.15+.
